src/intersections/boxes.jl

```diff
@@ -51,16 +51,16 @@ function intersecttype(f::Function, r::Ray{Dim,T}, b::Box{Dim,T}) where {Dim,T}
   lo, up = coordinates.(extrema(b))
   orig = coordinates(origin(r))
 
-  tmin = zero(T)
-  tmax = typemax(T)
+  tmin = zero(T) / oneunit(T)
+  tmax = typemax(T) / oneunit(T)
 
   # check for intersection with slabs along with each axis
   for i in 1:Dim
     imin = (lo[i] - orig[i]) * invdir[i]
     imax = (up[i] - orig[i]) * invdir[i]
 
     # swap variables if necessary
-    invdir[i] < zero(T) && ((imin, imax) = (imax, imin))
+    invdir[i] < zero(eltype(invdir)) && ((imin, imax) = (imax, imin))
```

Comment:
So invdir has the correct eltype even though we didn't use oneunit(T) right? I am still confused how units propagate. Before we can address this issue here, we need to discuss units in more depth in an issue I think. Mind opening an issue with a proposal that addresses the problem of unit propagation consistently across the codebase?

Reply:
@juliohm In this sample, invdir is the inverse of direction. If the direction is a dimensionless number (the dimensionless dimension is 1, such as Float64), the dimension of invdir is the reciprocal of 1, which is still 1, then the type of invdir is the same as the direction. But the agreement here is because of coincidence, because the reciprocal of 1 is still 1.

If the dimension of direction is millimeters, then the dimension of invdir should be the derivative of millimeters, which is mm^-1. At this time, the units of direction and invdir are different, so T can not be used to represent. To correctly indicate the type of invdir, typeof(invdir) should be used.

In all the previous codes, because only Float32 and Float64 are considered, this issue has not been carefully considered. And because 1 is a unit of Float, too many operations related to 1 are ignored and simplified. Therefore, the code before the constant with the unit will expose these problems.

I agree with you that we should open an issue to discuss this matter in detail and carefully before making a decision.